prqlc/prqlc/src/parser.rs

```diff
@@ -1,34 +1,159 @@
 use anyhow::Result;
-use std::collections::HashMap;
+use itertools::Itertools;
+use std::path::PathBuf;
+use std::{collections::HashMap, path::Path};
 
-use crate::utils::IdGenerator;
-use crate::{Errors, SourceTree};
-use prqlc_ast::stmt::Stmt;
+use crate::ast::{ModuleDef, Stmt, StmtKind};
+use crate::{Error, Errors, SourceTree, WithErrorInfo};
 
-pub fn parse(file_tree: &SourceTree<String>) -> Result<SourceTree<Vec<Stmt>>> {
-    let mut res = SourceTree::default();
-    let mut errors = Vec::new();
+pub fn parse(file_tree: &SourceTree) -> Result<ModuleDef> {
+    let source_files = linearize_tree(file_tree)?;
 
+    // reverse the id->file_path map
     let ids: HashMap<_, _> = file_tree.source_ids.iter().map(|(a, b)| (b, a)).collect();
-    let mut id_gen = IdGenerator::<usize>::new();
 
-    for (path, source) in &file_tree.sources {
+    // init the root module def
+    let mut root = ModuleDef {
+        name: "Project".to_string(),
+        stmts: Vec::new(),
+    };
+
+    // parse and insert into the root
+    let mut errors = Vec::new();
+    for source_file in source_files {
         let id = ids
-            .get(path)
+            .get(&source_file.file_path)
             .map(|x| **x)
-            .unwrap_or_else(|| id_gen.gen() as u16);
+            .expect("source tree has malformed ids");
 
-        match prqlc_parser::parse_source(source, id) {
+        match prqlc_parser::parse_source(source_file.content, id) {
             Ok(stmts) => {
-                res.sources.insert(path.clone(), stmts);
-                res.source_ids.insert(id, path.clone());
+                insert_stmts_at_path(&mut root, source_file.module_path, stmts);
             }
             Err(errs) => errors.extend(errs),
         }
     }
     if errors.is_empty() {
-        Ok(res)
+        Ok(root)
     } else {
         Err(Errors(errors).into())
     }
 }
+
+struct SourceFile<'a> {
+    file_path: &'a PathBuf,
+    module_path: Vec<String>,
+    content: &'a String,
+}
```

Comment:
Is it idiomatic to have refs to traditionally owned types like `Pathbuf` & `String`? I would have generally expected to see `&'a Path` xor `PathBuf`; same for string 

(but no _great_ stress obv...)

Reply:
https://github.com/PRQL/prql/pull/4222